File: backend/app/api/schemas/wallet.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field


class BalanceItemResponse(BaseModel):
    asset: str
    available: str


class BalanceListResponse(BaseModel):
    items: list[BalanceItemResponse] = Field(default_factory=list)


class TransactionItemResponse(BaseModel):
    id: UUID
    type: str
    status: str
    source_asset: str | None = None
    dest_asset: str | None = None
    amount: str
    created_at: datetime
    direction: str | None = None


class TransactionListResponse(BaseModel):
    total_items: int
    items: list[TransactionItemResponse] = Field(default_factory=list)


class ExchangeRequest(BaseModel):
    source_asset: str = Field(max_length=6)
    destination_asset: str = Field(max_length=6)
    amount: str


class WithdrawRequest(BaseModel):
    asset: str = Field(max_length=6)
    amount: str


class TransferRequest(BaseModel):
    email: EmailStr
    asset: str = Field(max_length=6)
    amount: str


class WalletMutationResponse(BaseModel):
    id: UUID
    type: str
    status: str = "COMPLETED"

File: frontend/src/types/wallet.ts
export type TransactionStatus =
  | 'submitted'
  | 'pending'
  | 'in_progress'
  | 'succeeded'
  | 'failed'

export type TransactionType = 'deposit' | 'withdrawal' | 'exchange' | 'transfer'

export interface TransactionStatusEvent {
  request_id: string
  status: TransactionStatus
  type: TransactionType
  error: string | null
}

export interface BalanceItem {
  asset: string
  amount: string
  locked: string
}

export type BalanceList = {
  items: BalanceItem[]
}

export interface TransactionItem {
  id: string
  request_id: string
  type: TransactionType
  status: TransactionStatus
  source_asset: string | null
  dest_asset: string | null
  amount: string
  error: string | null
  created_at: string
  updated_at: string
  direction?: 'in' | 'out'
}

export type TransactionList = {
  total_items: number
  items: TransactionItem[]
}

export type CurrencyItem = {
  label: string
  name: string
  type: string
  precision: number
}

export type UserReferenceItem = {
  user_id: string
  email: string
}

export type DataList<T> = {
  items: T[]
}

export type WalletMutationResponse = {
  id: string
  type: string
  status: string
}

export type ExchangeRequest = {
  source_asset: string
  destination_asset: string
  amount: string
}

export type WithdrawRequest = {
  asset: string
  amount: string
}

export type TransferRequest = {
  email: string
  asset: string
  amount: string
}
